### PR TITLE
Set cache-control headers to reduce server load (fixes #412)

### LINKED ISSUE #412: Fix missing cache headers
https://git.radicallyopensecurity.com/nlnet/off-ngid-lemmy/-/issues/10

### PR DESCRIPTION
**Pull request by Nutomic**
See code comment for explanation.

### CODE REVIEW

**Comment by SleeplessOne1917** [comment]
I don't see the code comment you're referring to in the changes.

**Comment by Nutomic** [comment]
Sorry I forgot to commit some changes, check again now.

**Review by SleeplessOne1917 — APPROVED** [review]
There's a small change to make, but it's not consequential enough to mark as request changes.

**Comment by alectrocute** [comment]
I’m not a fan of all the middleware being in the same file. We should keep it a folder, yes?

**Comment by alectrocute** [comment]
Does this resolve https://github.com/LemmyNet/lemmy-ui/issues/1375?

File: src/server/middleware.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  res,
+  next,
+}: {
+  res: Response;
+  next: NextFunction;
+}) {
+  const user = UserService.Instance;
+  let caching;
+  if (user.auth()) {
+    caching = "private";
```
[the remark this answers is not in the document]
Comment (pixlguru) [reply src/server/middleware.ts:35]: Are these really necessary?

`unsafe-inline` and `unsafe-eval`  introduce serious security risks.

Please avoid usage or use with nonce

File: src/server/middleware.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+export function setDefaultCsp({
+  res,
+  next,
+}: {
+  res: Response;
+  next: NextFunction;
+}) {
+  res.setHeader(
+    "Content-Security-Policy",
+    `default-src 'self'; manifest-src *; connect-src *; img-src * data:; script-src 'self' 'unsafe-inline' 'unsafe-eval'; style-src 'self' 'unsafe-inline'; form-action 'self'; base-uri 'self'; frame-src *; media-src *`
```
Comment (pixlguru) [inline src/server/middleware.ts:13]: Are these really necessary?

`unsafe-inline` and `unsafe-eval` introduce serious security risks.

Please avoid usage or use with nonce

Comment (Nutomic) [reply src/server/middleware.ts:13]: Afaik they are required by the js framework, would be great if we could get rid of them.

File: src/server/middleware.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (Nutomic) [reply src/server/middleware.ts:35]: That will cause problems if you write a post, then go to the post listing which is cached, and is missing your new post. In any case the main server load is probably from unauthenticated users so this should already help a lot.

File: src/server/middleware.ts (hunk 2 above)
Comment (dessalines) [reply src/server/middleware.ts:13]: @pixlguru you can try removing these as a separate PR, but make sure you do real testing on it.

Comment (pixlguru) [reply src/server/middleware.ts:13]: Maybe I could do this after 18.1 is out, if I remember correctly ,last time I took a look unsave-eval was used by websocket code, so there is a chance that removing it has no side-effects.

**Pull request merged**